From owner-linux-pci@atrey.karlin.mff.cuni.cz Wed Aug  2 10:36:50 2006
From: Daniel Ritz <daniel.ritz-ml@swissonline.ch>
To: Greg KH <gregkh@suse.de>, Andrew Morton <akpm@osdl.org>
Subject: PCI: use PCI_BIOS as last fallback
Date: Wed, 2 Aug 2006 19:36:40 +0200
Cc: Marcus Better <marcus@better.se>
Content-Disposition: inline
Message-Id: <200608021936.41102.daniel.ritz-ml@swissonline.ch>

[PATCH] PCI: use PCI_BIOS as last fallback

there was a change in 2.6.17 which affected the order in which the PCI access
methods are probed. this gives regressions on some machines with broken BIOS.
the problem is that PCI_BIOS sometimes reports the last bus wrong, leaving cardbus
non-funcational. previously those system worked fine with direct access.
fix it by chaning the order of the probing, having PCI_BIOS as the last fallback.

Signed-off-by: Daniel Ritz <daniel.ritz@gmx.ch>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/i386/pci/init.c |   10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

--- gregkh-2.6.orig/arch/i386/pci/init.c
+++ gregkh-2.6/arch/i386/pci/init.c
@@ -11,13 +11,13 @@ static __init int pci_access_init(void)
 #endif
 	if (raw_pci_ops)
 		return 0;
-#ifdef CONFIG_PCI_BIOS
-	pci_pcbios_init();
-#endif
-	if (raw_pci_ops)
-		return 0;
 #ifdef CONFIG_PCI_DIRECT
 	pci_direct_init();
+	if (raw_pci_ops)
+		return 0;
+#endif
+#ifdef CONFIG_PCI_BIOS
+	pci_pcbios_init();
 #endif
 	return 0;
 }
